RippleCarryAdder/Tests.qs

```diff
@@ -235,16 +242,16 @@ namespace Quantum.Kata.RippleCarryAdder {
     operation T17_ArbitraryAdder_Test () : Unit {
         // 4 bits seems reasonable - any more than that will take forever
         for (i in 1 .. 4){
-            let testOp = QubitArrayAdderWrapper(i, ArbitraryAdder, _);
-            AssertOperationImplementsBinaryFunction(testOp, BinaryAdder(_, i), 2 * i, i + 1);
+			let testOp1 = QubitArrayAdderWrapper(i, ArbitraryAdder, _);
+
+            let testOp2 = QubitArrayAllocateAdderWrapper(i, ArbitraryAdder, _);
+			let refOp = QubitArrayAllocateAdderWrapper(i, ArbitraryAdder_Reference, _);
+
+			AssertOperationImplementsBinaryFunction(testOp1, BinaryAdder(_, i), 2 * i, i + 1);
+            AssertOperationsEqualReferenced(i, testOp2, refOp);
```

Comment:
The way this check is written now doesn't actually test the solution, though it took me a while to figure out why exactly.

`AssertOperationsEqualReferenced` compares two operations assuming that they are operations on `i` qubits: it prepares `i` qubits in a certain entangled state, passes them to the first operation and then to the adjoint of the second operation and then checks that these qubits ended in the same state as they started.  `QubitArrayAllocateAdderWrapper` wraps the operations we want to compare, but it doesn't actually use the qubits passed to them as the last parameter, it allocates fresh qubits for all parameters and runs the operations on them. This means that it has no effect on the qubits passed to it, and any operation will be equal to any other operation. Indeed, if you comment out `AssertOperationImplementsBinaryFunction` line above, you'll see that any operation, including empty one, will pass the test!

---

I did some experimenting with the code to see how to get the behavior we need, and it is a bit trickier than I expected when I was writing up the issue, so let me go into more details here.

We need to compare two operations which take parameters `a, b, sum, carry` each; of these parameters `a, b, carry` are allowed to be in arbitrary state, and `sum` is guaranteed to be in 0 state. We can do this using the following modification of the [`AssertOperationsEqualReferenced` implementation](https://github.com/microsoft/qsharp-runtime/blob/master/src/Simulation/Intrinsic/Diagnostics/AssertOperationsEqualReferenced.qs):

    using ((reference, target, sum) = (Qubit[2*i+1], Qubit[2*i+1], Qubit[i])) {
        let a = target[0..i-1];
        let b = target[i..2*i-1];
        let carry = Tail(target);

        _prepareEntangledState(reference, target);
        ArbitraryAdder(a, b, sum, carry);
        Adjoint ArbitraryAdder_Reference(a, b, sum, carry);
        Adjoint _prepareEntangledState(reference, target);
        AssertAllZero(reference + target);
    }

You can see that the parameters that accept arbitrary states are passed to `_prepareEntangledState` following the same verification logic as is used for unitary comparison, and the parameter that has to start in 0 state is left alone with only the operations applied to it.